checker/src/call_graph.rs

```diff
@@ -0,0 +1,936 @@
+            Entry::Vacant(v) => {
+                let node_idx = self.graph.add_node(CallGraphNode::new_root(defid));
+                *v.insert(node_idx)
+            }
+        }
+    }
+
+    /// Add a dominance relationship to the call graph.
+    /// Denotes that `defid1` is dominated by `defid2`.
+    pub fn add_dom(&mut self, defid1: DefId, defid2: DefId) {
+        let node_idx1 = self.get_or_insert_node(defid1);
+        let node_idx2 = self.get_or_insert_node(defid2);
+        self.dominance
+            .entry(node_idx1)
+            .or_insert_with(HashSet::<NodeIdx>::new)
+            .insert(node_idx2);
+    }
+
+    /// Add a new RType to the call graph's `rtypes`.
+    fn add_rtype(&mut self, rtype_str: Box<str>) -> RTypeIdx {
+        let new_rtype_id = self.rtypes.len() as RTypeIdx;
+        match self.rtypes.entry(rtype_str.to_owned()) {
+            Entry::Occupied(rtype) => rtype.get().id,
+            Entry::Vacant(v) => v.insert(RType::new(new_rtype_id, rtype_str)).id,
+        }
+    }
+
+    /// Add a new edge to the call graph.
+    /// The edge is a call edge from `caller_id` to `callee_id` with type `rtype_str`.
+    pub fn add_edge(&mut self, caller_id: DefId, callee_id: DefId, rtype_str: Box<str>) {
+        let rtype_id = self.add_rtype(rtype_str);
+        let caller_node = self.get_or_insert_node(caller_id);
+        let callee_node = self.get_or_insert_node(callee_id);
+        let mut existing_rtypes = HashSet::<RTypeIdx>::new();
+        for edge in self.graph.edges_connecting(caller_node, callee_node) {
+            existing_rtypes.insert(edge.weight().rtype_id);
+        }
+        if !existing_rtypes.contains(&rtype_id) {
+            self.graph
+                .add_edge(caller_node, callee_node, CallGraphEdge::new(rtype_id));
+        }
+    }
+
+    /// Find a node in the call graph given a `name` that may appear as
+    /// a substring within the node's name. The first such node is returned, if any.
+    fn get_node_by_name(&self, name: &str) -> Option<NodeIdx> {
+        for node_idx in self.graph.node_indices() {
+            if let Some(node) = self.graph.node_weight(node_idx) {
+                if node.name.contains(name) {
+                    return Some(node_idx);
+                }
+            }
+        }
+        None
+    }
+
+    /// Deduplicate edges by only including a single edge for for each
+    /// (caller, callee) pair that has an edge.
+    ///
+    /// If there are multiple edges from a caller to a callee with different types,
+    /// this has the effect of including only one of those edges.
+    fn deduplicate_edges(&self) -> CallGraph {
+        let mut edges = HashSet::<(NodeIdx, NodeIdx)>::new();
+        let graph = self.graph.filter_map(
+            |_, node| Some(node.to_owned()),
+            |edge_idx, edge| {
+                self.graph.edge_endpoints(edge_idx).and_then(|endpoints| {
+                    if edges.contains(&endpoints) {
```

Comment:
Instead of doing a check if the key there and then adding it if not, just insert the key every time and use the boolean result to determine if the key has been seen before.

Reply:
Good idea, I'll change that